src/unicode/capi.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * This file is part of the "libunicode" project
+ *   Copyright (c) 2021 Christian Parpart <christian@parpart.family>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if !defined(__cplusplus)
+extern "C" {
+#endif
+
+/// UTF-8 character or 8bit segment of an UTF-8 character.
+typedef char u8_char_t;
+
+/// UTF-32 codepoint.
+typedef uint32_t u32_char_t;
```

Comment:
This type might need a bit more explanation, e.g:
- Endianess? (prolly architecture bound I guess, or UTF-32LE vs. BE)
- its value range: unicode only defines 0 - 0x10FFFF, what happens to the higher bits? Left untouched (thus guaranteed not to be set/changed by the lib)?
- other convenience stuff, like validators or converters...

Reply:
Hm, at least that one I wanted to add, too: `char32_t` is behind the scenes a `uint32_least_t`, so I should match that, too.

> LE vs BE

I agree, that's not directly clear. I think one should at least document that this is the endianess of the underlying architecture this library is compiled and deployed on.

> Unicode range (0 - 0x10FFFF)

Oh I didn't count that in, right. I think it should at least be made clear how the functions using these types behave, i.e. boolean functions that test on a property but receive an invalid input do return false, etc. 

> other convenience stuff, like validators or converters

converters between UTF-8 <-> 32 I can provide. many thanks. Validators, umm... anything you'd prefer to see in such a library (in or not in the context of terminal emus / apps)?

Reply:
Ah not sure what your want to provide here, at least a simple `is_valid_codepoint(u32_char_t codepoint)` might help peeps not to bother with real unicode internals and such.

And claiming what bits are used by the lib might help hackers to overload higher bits with their own meaning, which is a quite common trick to pack information along with codepoints, I think several text processors/emulators do this (we at least do it in xterm.js). But for that it would be needed, that the lib guarantees to only manipulate/take into account unicode bits. Not sure if thats a viable goal though.

Reply:
Nice point. I will provide some making macros and bit width definitions to help support those ideas

Reply:
Well we use it as the main text content of a cell, and put the wcwidth value and `isCombining` next to it:

https://github.com/xtermjs/xterm.js/blob/96806ee384a57cbf4d5192e86aa5dbe437a313af/src/common/buffer/BufferLine.ts#L13-L17